RecoMuon/TrackerSeedGenerator/plugins/TSGForOIFromL2.cc

```diff
@@ -0,0 +1,474 @@
+
+  // Stepping Helix Propagator for propogation from muon system to tracker
+  edm::ESHandle<Propagator> SHPOpposite;
+  iSetup.get<TrackingComponentsRecord>().get("hltESPSteppingHelixPropagatorOpposite", SHPOpposite);
+
+  // Loop over the L2's and make seeds for all of them
+  LogTrace(theCategory) << "TSGForOIFromL2::produce: Number of L2's: " << l2TrackCol->size();
+  for (unsigned int l2TrackColIndex(0); l2TrackColIndex!=l2TrackCol->size(); ++l2TrackColIndex) {
+
+    const reco::TrackRef l2(l2TrackCol, l2TrackColIndex);
+
+    // Container of Seeds
+    std::vector<TrajectorySeed> out;
+    LogTrace("TSGForOIFromL2") << "TSGForOIFromL2::produce: L2 muon pT, eta, phi --> " << l2->pt() << " , " << l2->eta() << " , " << l2->phi() << endl;
+ 
+    FreeTrajectoryState fts = trajectoryStateTransform::initialFreeState(*l2, magfieldH.product());
+
+    dummyPlane->move(fts.position() - dummyPlane->position());
+    TrajectoryStateOnSurface tsosAtIP = TrajectoryStateOnSurface(fts, *dummyPlane);
+    LogTrace("TSGForOIFromL2") << "TSGForOIFromL2::produce: Created TSOSatIP: " << tsosAtIP << std::endl;
+    
+    // Get the TSOS on the innermost layer of the L2
+    TrajectoryStateOnSurface tsosAtMuonSystem = trajectoryStateTransform::innerStateOnSurface(*l2, *geometryH, magfieldH.product());
+    LogTrace("TSGForOIFromL2") << "TSGForOIFromL2::produce: Created TSOSatMuonSystem: " << tsosAtMuonSystem <<endl;
+    
+    LogTrace("TSGForOIFromL2") << "TSGForOIFromL2::produce: Check the error of the L2 parameter and use hit seeds if big errors" << endl;
+
+    StateOnTrackerBound fromInside(propagatorAlong.get());
+    TrajectoryStateOnSurface outerTkStateInside = fromInside(fts);
+
+    StateOnTrackerBound fromOutside(&*SHPOpposite);
+    TrajectoryStateOnSurface outerTkStateOutside = fromOutside(tsosAtMuonSystem);
+
+    // Check if the two positions (using updated and not-updated TSOS) agree withing certain extent.
+    // If both TSOSs agree, use only the one at vertex, as it uses more information. If they do not agree, search for seeds based on both.
+    bool useBoth = false;
+    if (outerTkStateInside.isValid() && outerTkStateOutside.isValid()) {
+      //following commented out variables dist1 (5 par compatibility of tsos at outertracker surface)  
+      //dist2 (angle between two tsos) could further be explored in combination of L2 valid hits for seeding. So kept for 
+      //future developers 
+      //auto dist1 = match_Chi2(outerTkStateInside,outerTkStateOutside);//for future developers
+      //auto dist2 = deltaR(outerTkStateInside.globalMomentum(),outerTkStateOutside.globalMomentum());//for future developers
+      //if ((dist1 > tsosDiff1_ || dist2 > tsosDiff2_) && l2->numberOfValidHits() < 20) useBoth = true;//for future developers
+      if (l2->numberOfValidHits() < numL2ValidHitsCutAllEta_) useBoth = true;
+      if (l2->numberOfValidHits() < numL2ValidHitsCutAllEndcap_ && (std::abs(l2->eta()) > eta7_)) useBoth = true;
```

Comment:
```std::abs(l2->eta())``` is recomputed up to 10 times at every iteration.
Maybe one could compute it only once (right after L128?), and use that value everywhere